File: Server/src/main/java/core/game/interaction/inter/QuestTabInterface.java
package core.game.interaction.inter;

import core.game.component.Component;
import core.game.component.ComponentDefinition;
import core.game.component.ComponentPlugin;
import core.game.node.entity.player.Player;
import core.game.node.entity.player.link.diary.AchievementDiary;
import core.game.node.entity.player.link.diary.DiaryType;
import core.game.node.entity.player.link.quest.Quest;
import core.plugin.Initializable;
import core.plugin.Plugin;

/**
 * Handles the quest tab reward buttons.
 * @author Emperor
 * @author Vexia
 */
@Initializable
public class QuestTabInterface extends ComponentPlugin {

	@Override
	public Plugin<Object> newInstance(Object arg) throws Throwable {
		ComponentDefinition.put(274, this); // Quests
		ComponentDefinition.put(259, this); // Achievement diary
		return this;
	}

	@Override
	public boolean handle(Player p, Component component, int opcode, int button, int slot, int itemId) {
		p.getPulseManager().clear();
		switch (component.getId()) {
			case 274:
//				if (!GameWorld.isEconomyWorld()) {
//					p.getSavedData().getSpawnData().handleButton(p, button);
//				}
				switch (button) {
					case 3:
						p.getAchievementDiaryManager().openTab();
						return true;
					case 10:
						break;
					default:
//						if (GameWorld.isEconomyWorld()) {
							//System.out.println("Quest button: " + button);
							Quest quest = p.getQuestRepository().forButtonId(button);
							if (quest != null) {
								p.getInterfaceManager().open(new Component(275));
								quest.drawJournal(p, quest.getStage(p));
								return true;
							}
//						}
						return false;
				}
				break;
			case 259:
				switch (button) {
					case 8:
						p.getInterfaceManager().openTab(2, new Component(274));
						return true;
					default:
						AchievementDiary diary = p.getAchievementDiaryManager().getDiary(DiaryType.forChild(button));
						if (diary != null) {
							diary.open(p);
						}
						return true;
				}
		}
		return true;
	}

}

File: Server/src/main/content/global/handlers/iface/tabs/QuestTabInterface.kt
package content.global.handlers.iface.tabs

import content.global.handlers.iface.tabs.QuestTabUtils.showRequirementsInterface
import core.api.openInterface
import core.game.component.Component
import core.game.interaction.InterfaceListener
import core.game.node.entity.player.link.diary.DiaryType
import org.rs09.consts.Components

class QuestTabInterface : InterfaceListener {
    override fun defineInterfaceListeners() {
        on(Components.QUESTJOURNAL_V2_274) { player, _, _, buttonID, _, _ ->
            if (buttonID == 3) {
                player.achievementDiaryManager.openTab()
            } else {
                val quest = player.questRepository.forButtonId(buttonID)
                if (quest != null) {
                    openInterface(player, Components.QUESTJOURNAL_SCROLL_275)
                    quest.drawJournal(player, quest.getStage(player))
                } else {
                    showRequirementsInterface(player, buttonID)
                }
            }
            return@on true
        }

        on(Components.AREA_TASK_259) { player, _, _, buttonID, _, _ ->
            if (buttonID == 8) {
                player.interfaceManager.openTab(2, Component(Components.QUESTJOURNAL_V2_274))
            } else {
                player.achievementDiaryManager.getDiary(DiaryType.forChild(buttonID))?.open(player)
            }
            return@on true
        }
    }
}